src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +150,131 @@ private void logTargetPath() {
+        sortCamerasByLatestResultTimestamp(newResultCameras);
+
+        updateEstimatedPoseFromVision(newResultCameras);
+    }
+
+    private void updateCameras() {
+        for (AprilTagCamera aprilTagCamera : aprilTagCameras)
+            aprilTagCamera.update();
+    }
+
+    private AprilTagCamera[] getCamerasWithResults() {
+        AprilTagCamera[] camerasWithResults = new AprilTagCamera[aprilTagCameras.length];
+        int index = 0;
+
+        for (AprilTagCamera aprilTagCamera : aprilTagCameras)
+            if (aprilTagCamera.hasNewResult()) {
+                camerasWithResults[index] = aprilTagCamera;
+                index++;
+            }
+
+        return Arrays.copyOf(camerasWithResults, index);
     }
 
-    private List<PoseEstimator6328.VisionObservation> getViableVisionObservations() {
-        List<PoseEstimator6328.VisionObservation> viableVisionObservations = new ArrayList<>();
+    private void sortCamerasByLatestResultTimestamp(AprilTagCamera[] aprilTagCameras) {
+        QuickSort.sort(aprilTagCameras, AprilTagCamera::getLatestResultTimestampSeconds);
+    }
+
+    /**
+     * Updates the estimated pose from each camera, each at the timestamp of their latest result.
+     */
+    private void updateEstimatedPoseFromVision(AprilTagCamera[] aprilTagCameras) {
         for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
-            final PoseEstimator6328.VisionObservation visionObservation = getVisionObservation(aprilTagCamera);
-            if (visionObservation != null)
-                viableVisionObservations.add(visionObservation);
+            addVisionObservation(
+                    aprilTagCamera.getEstimatedRobotPose(),
+                    aprilTagCamera.calculateStandardDeviations(),
+                    aprilTagCamera.getLatestResultTimestampSeconds()
+            );
         }
-        return viableVisionObservations;
     }
 
-    private PoseEstimator6328.VisionObservation getVisionObservation(AprilTagCamera aprilTagCamera) {
-        aprilTagCamera.update();
-        if (!aprilTagCamera.hasNewResult())
-            return null;
-        final Pose2d robotPose = aprilTagCamera.getEstimatedRobotPose();
-        if (robotPose == null || robotPose.getTranslation() == null || robotPose.getRotation() == null)
-            return null;
+    private void addVisionObservation(Pose2d estimatedPose, StandardDeviations standardDeviations, double timestamp) {
+        if (isVisionObservationTooOld(timestamp))
+            return;
+
+        final Pose2d previousEstimatedPoseAtObservationTimestamp = getPoseAtTimestamp(timestamp);
+        if (previousEstimatedPoseAtObservationTimestamp == null)
+            return;
+
+        final Transform2d odometryPoseToObservationTimestampPoseTransform = new Transform2d(odometryPose, previousEstimatedPoseAtObservationTimestamp);
+        final Pose2d newEstimatedPoseAtObservationTime = estimatedPose.plus(odometryPoseToObservationTimestampPoseTransform);
```

Comment:
1. I can see where you got confused from 6328's logic, but the logic here is incorrect and dare I say not understandable.
2. While everything you programmed here and in this PR is beautiful and very cool, I'd like you to try and think more about the logic you write and understand it, rather than copy it without batting an eye (at least, that's what seems to happen here).